### PR TITLE
Add feature to ignore git commits by id

### PR DESCRIPTION
**Pull request by philippn**
Hey @mathieucarbou ,

here is the PR, thanks for your consideration.

Kind regards, 
Philipp

### CODE REVIEW

**Comment by philippn** [comment]
I also rebased it onto master just now

**Comment by mathieucarbou** [comment]
@philippn : thanks a lot! Could you please also update the `README.md` in the git module to add a few lines to document this new feature ? There is a usage example that can be updated at least, plus eventually adding a comment line to explain the setting or a line below or above.

@dbwiddis @hazendaz : PR looks good to me as-is but please add your review if you can since you contribute a lot you might have your word on it too ;-)

File: license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitPropertiesProvider.java
Comment (dbwiddis) [inline license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitPropertiesProvider.java:36]: This name should include `_KEY` in the name to distinguish it from the value.

File: license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitLookup.java
Comment (dbwiddis) [inline license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitLookup.java:57]: Why store as a `String` rather than as a `RevCommit`?  You can use `fromString()` to convert.  This buys you:
 - Case insensitivity on the hex strings since you're converting to numeric values
 - Type safety/self-documenting code
 - Automatic format checking (e.g., invalid commit id strings will throw IllegalArgumentException and let the user know which commit ID is problematic). Otherwise you'd fail silently.
 - Less space. The strings take 80 bytes, the `RevCommit` stores in a 20-byte array.
 - There's already an `equals()` based on the numeric value so you can compare the `RevCommit` directly (`commitsToIgnore.contains(commit)`) as you iterate rather than filling up the heap using `.name()` to create a temporary `String` for every commit, for every file, while iterating.

File: license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitLookup.java
Comment (dbwiddis) [inline license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitLookup.java:69]: Probably should explain what is expected in this argument.

File: license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitPropertiesProvider.java
Comment (dbwiddis) [inline license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitPropertiesProvider.java:20]: See above comment on wildcards.

File: license-maven-plugin-git/src/test/java/com/mycila/maven/plugin/license/git/GitLookupTest.java
Comment (dbwiddis) [inline license-maven-plugin-git/src/test/java/com/mycila/maven/plugin/license/git/GitLookupTest.java:164]: Unnecessary (as are the other changes in this file) if you use an overloaded constructor.

File: license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitLookup.java
Comment (dbwiddis) [inline license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitLookup.java:104]: This doesn't need to change if you use an overload. Also, interestingly, there is no note on time zones. 😁

**Review by dbwiddis — CHANGES_REQUESTED** [review]
Comments inline. 
 - [x] Also you should add a test to `GitLookupTest` to verify that it works.

File: license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitLookup.java
Comment (mathieucarbou) [reply license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitLookup.java:37]: @dbwiddis : I think the priority could be to add the checkstyle plugin to the project in order to help contributors :-)
Some IDE are configured by default to use wildcards when optimizing imports.

File: license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitPropertiesProvider.java
Comment (dbwiddis) [inline license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitPropertiesProvider.java:71]: One other suggestion.  Splitting on a comma forces users into a long one-line string here.  You could split on a non-hex digit, e.g., `[^0-9a-fA-F]+` (and also filter out empty strings after the split) which would allow flexibility such as listing commits on newlines like:
```xml
<license.git.commitsToIgnore>
  6eb93dcfc014bfd3980c1f1c43f10fd59422107e,
  ad1e59a6afc5a1a7edbdc8bcf949bb2f8682f074,
  70403fdbc424e5d2045389e685f9bb9bd696d5cd
</license.git.commitsToIgnore>
```

File: license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitLookup.java
Comment (dbwiddis) [reply license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitLookup.java:74]: Looking again, this is fine as-is.  Other than in tests, `GitLookup` is set up to be used by the `GitPropertiesProvider` with lazy initialization of a static variable, so there really should be only one entry point.

Comment (mathieucarbou) [reply license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitLookup.java:74]: Let's @philippn decide because both ways are debatable and ok IMO.

File: license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitPropertiesProvider.java
Comment (dbwiddis) [reply license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitPropertiesProvider.java:71]: By the way, the reason I'm making this suggestion is that I really think this is a cool feature and I'm eager to use it, and somewhat biased by how I want to use it. ;)

**Comment by philippn** [comment]
I'm a bit overwhelmed by all this feedback, please bear with me :-)

So in the order of the points being mentioned:

- I added a paragraph to the readme.md
- I added myself to the contributors, thanks for pointing that out
- I added two new tests to `GitLookupTest`; no here is actually where things start to get interesting. On my machine (Windows, IntelliJ) the tests weren't picked up, neither by IntelliJ itself nor by Maven. Dunno whats going on here but changing line 60 to `gitRepoRoot = unzipDestination.resolve("git-test-repo");` solved the issue for me, at least it it then running in IntelliJ.

Regarding the concrete things pointed out by @dbwiddis I will comment on each individually.

File: license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitLookup.java
Comment (philippn) [reply license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitLookup.java]: Roger, I changed that back

File: license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitLookup.java
Comment (philippn) [reply license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitLookup.java:74]: Since it is really only used at one place, lets leave it that way.

File: license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitPropertiesProvider.java
Comment (philippn) [reply license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitPropertiesProvider.java]: Roger, thanks for pointing that out

File: license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitLookup.java
Comment (philippn) [reply license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitLookup.java]: Nice catch, I have changed it accordingly.

File: license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitLookup.java
Comment (philippn) [reply license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitLookup.java]: Roger

File: license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitPropertiesProvider.java
Comment (philippn) [reply license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitPropertiesProvider.java]: Roger, I changed that back

File: license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitPropertiesProvider.java
Comment (philippn) [reply license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitPropertiesProvider.java:77]: I looked into this, I'm a bit vary to use lots of regex magic here as it is not easily testable. It would probably make sense to create a test for `GitPropertiesProvider`, but that would require factoring out the instantiation of the `GitLookup` into some kind of factory.

So for the time being, I added a `trim` operation to the streaming expression, that should basically result in the same behaviour. So technically, I believe your example above should work :-)

Comment (dbwiddis) [reply license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitPropertiesProvider.java:77]: > I looked into this, I'm a bit vary to use lots of regex magic here as it is not easily testable. 

Fair enough. I hit regex101 to play around with my proposal and it took a few iterations to confirm what I expected; and the added complexity of producing empty strings that you then have to filter was undesirable. I like your `trim()` solution and it (should) work with my use case.

> It would probably make sense to create a test for `GitPropertiesProvider`, but that would require factoring out the instantiation of the `GitLookup` into some kind of factory.

Yeah. I'm not fond of the existing lazy initialization but haven't had time to think of a better way to do it.

**Review by dbwiddis — APPROVED** [review]

**Comment by dbwiddis** [comment]
> I'm a bit overwhelmed by all this feedback, please bear with me :-)

Speaking for project maintainers in general, we are thrilled with contributions from the community and we will happily bear with you as needed.  Thanks for contributing and being so gracious with the feedback.

This LGTM at this point.   Not sure what's going on with the tests, but if @mathieucarbou wants to merge this push out an rc3 that I can fiddle with real-life testing on my own project I'm happy to do that!

**Comment by dbwiddis** [comment]
@hazendaz JDK 17 / 18 tests are failing with "Not executing Javadoc as the project is not a Java classpath-capable package" ... hope that's something you're already working on. Do we need to rebase this PR?

**Comment by hazendaz** [comment]
> are failing with "Not executing Javadoc as the project is not a Java classpath-capable package" ... hope that's something you're already working on. Do we need to rebase this PR?

Had not seen that yet but will take a look at that one.  Thanks.

**Comment by hazendaz** [comment]
@dbwiddis javadoc may be unrelated issue.  The issue that jump sout at failure is test coverage drops down by 1% on 17/18.  Not sure exactly why but I had the same issue when setting jacoco up there.  Typically I go with more strict enforcement of coverage but in this case wanted to go with the closest to IDE coverage stats in Eclipse.  So to fix that @philippn, its easy go to the 2 submodule poms (license-maven-plugin and license-maven-plugin-git.  Just drop the jacoco.minimum coverage number by 1% in both.  That should get it to green here.

**Review by mathieucarbou — DISMISSED** [review]
Please wait for #339 to be merged first and rebase onto after.

File: license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitPropertiesProvider.java
Comment (mathieucarbou) [reply license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitPropertiesProvider.java:77]: > > It would probably make sense to create a test for `GitPropertiesProvider`, but that would require factoring out the instantiation of the `GitLookup` into some kind of factory.
> 
> Yeah. I'm not fond of the existing lazy initialization but haven't had time to think of a better way to do it.

So that is fixed in #339

**Comment by mathieucarbou** [comment]
> Please wait for #339 to be merged first and rebase onto after.

#339 is merged!

**Review by dbwiddis — APPROVED** [review]

**Comment by mathieucarbou** [comment]
@dbwiddis : your merge commit looks weird ;-)
The correct way for this PR is to rebase the original commit on top of master, solve conflicts, and merge it with a merge commit.
Don't know if @philippn has the time for that, or otherwise if not we rebase and force push his commit if his branch his opened, or create a new PR.

**Review by mathieucarbou — CHANGES_REQUESTED** [review]
PR has to be rebased on top of master and conflicts solved as part of the rebase.

**Comment by dbwiddis** [comment]
> @dbwiddis : your merge commit looks weird ;-) 

Github commandline ain't all it's cracked up to be.

Unfortunately the whole #339 refactor made resolving conflicts nontrivial.  Will try again.

**Comment by mathieucarbou** [comment]
> Don't ask why there are two merge commits. But I've fixed conflicts 4 times now and I think it's correct now. 😁

I'll check. Not normal... gît fetch upstream && gît rebase -i upstream/master then fix conflits during rebase, there should be one commit and no merge commit

**Comment by dbwiddis** [comment]
Yeah, I know how it's *supposed* to work.  Between the GitHub CLI and my IDE things didn't work out well the first time.

I successfully merged the second go around but then ended up unable to push back to this branch, so I cherry-picked the merge commit onto this branch and it seemed to work, with the exception of showing two commits.

**Comment by mathieucarbou** [comment]
> 

That is normal that you were not able to push back because rebasing changes the history and commit ID to place it on top of master. You have to force the puah, which will rewrite the branch history for this commit and GitHub will show your icon next to @philippn 's icon

**Comment by dbwiddis** [comment]
Well, in any case, the two commits are identical and the result of all this is correctly working code that can be merged.   If someone wants to put in the work to back those commits out and start afresh, feel free.  Having done the same changes 4 times, I'm done. :grin:

**Comment by dbwiddis** [comment]
Well, that worked just fine at home on my mac.  😏

**Comment by mathieucarbou** [comment]
> Well, that worked just fine at home on my mac. smirk

Still not rebased on master if we look at the graph: https://github.com/mathieucarbou/license-maven-plugin/network

File: license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitLookup.java
Comment (mathieucarbou) [inline license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitLookup.java:107]: there is a bug here: we need to filter out empty strings after the split

**Review by mathieucarbou — CHANGES_REQUESTED** [review]

**Review by mathieucarbou — APPROVED** [review]

File: license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitLookup.java
Comment (mathieucarbou) [reply license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/GitLookup.java]: @dbwiddis @philippn : FYI added a second commit to rework that part

**Pull request merged**
